xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/Visualization.h

```diff
@@ -145,36 +351,311 @@ namespace addon
+    /// @param[in] songName             The name of the currently-playing song
+    ///
     virtual bool Start(int channels, int samplesPerSec, int bitsPerSample, std::string songName) { return true; }
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @ingroup cpp_kodi_addon_visualization
+    /// @brief Used to inform the visualization that the rendering control was
+    /// stopped
+    ///
     virtual void Stop() {}
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @ingroup cpp_kodi_addon_visualization
+    /// @brief Pass audio data to the visualization
+    ///
+    /// @param[in] audioData            The raw audio data
+    /// @param[in] audioDataLength      Length of the audioData array
+    /// @param[in] freqData             The [FFT](https://en.wikipedia.org/wiki/Fast_Fourier_transform)
+    ///                                 of the audio data
+    /// @param[in] freqDataLength       Length of frequency data array
+    ///
+    /// Values **freqData** and **freqDataLength** are used if GetInfo() returns
+    /// true for the `wantsFreq` parameter. Otherwise, **freqData** is set to
+    /// `nullptr` and **freqDataLength** is `0`.
+    ///
     virtual void AudioData(const float* audioData, int audioDataLength, float* freqData, int freqDataLength) {}
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @ingroup cpp_kodi_addon_visualization
+    /// @brief Used to inform Kodi that the rendered region is dirty and need an
+    /// update
+    ///
+    /// @return True if dirty
+    ///
     virtual bool IsDirty() { return true; }
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @ingroup cpp_kodi_addon_visualization
+    /// @brief Used to indicate when the add-on should render
+    ///
     virtual void Render() {}
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @ingroup cpp_kodi_addon_visualization
+    /// @brief Used to get the number of buffers from the current visualization
+    ///
+    /// @param[out] wantsFreq           Indicates whether the add-on wants FFT
+    ///                                 data. If set to true, the **freqData**
+    ///                                 and **freqDataLength** parameters of
+    ///                                 AudioData() are used.
+    /// @param[out] syncDelay           Number of buffers from the current !!! @todo
```

Comment:
Comes from there https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIVisualisationControl.cpp#L253-L266

The `m_numBuffers` is set by them here.
Seen also a bad part there, seems like the length is always the same, otherwise there would have already been a crash. There becomes the length from another data passed to the addon.

Need a change there.